### DISCUSSION TITLE
Plugin Store Design Decisions

### DISCUSSION THREAD

**Discussion by webcrumbs-community** [webcrumbs-community/webcrumbs#20] [Ideas]
A conversation about the design and architecture of the plugin store.

**Comment by webcrumbs-community** [comment]
In term of stack, we're thinking of...
- Nodejs + Reactjs + Nextjs
- Webpack + Dynamic Imports

We'll need to store Webpack bundles in a **reliable place** for safety reasons. Endpoints like registry.webcrumbs.org.
We should allow developers to **access the source code** of third party plugins to scrutinize safety issues.
We may use ⭐️⭐️⭐️⭐️⭐️ , installation numbers and other **social proof** elements to allow the community to **"darwin select"** the best.

**Comment by webcrumbs-community** [comment]
A few principles:

1. **Zero coupling between plugins** (no imports between plugins, no shared state, although shared libraries through module federation is recommended, for performance reasons).
2. **Near zero coupling between end website and plugins** (end website shouldn't assume the plugin work with a particular framework; any necessary communications are done with callbacks and simple events)
3. **CSS from one plugin should not affect the other** (although they should be able to use the main design system of the end website, and allow customization of specific styles through the admin panel)
4. **End website is free to decide which version of plugins to use** (a new plugin version should not break end websites using it)

**Reply by ghana7989** [reply]
What if the plugin needs to access the state of the application?
for example if the plugin wanna display hello ${username} how can we do that?

**Reply by webcrumbs-community** [reply]
@ghana7989 I guess this would fall into the **near** zero coupling between the website and the plugins. The website can pass props to the plugins. For safety reasons, we may think of a way of getting the users permission if the data the plugin want to access exceeds the basic. Kind of like Google oAuth, that will ask users if they are sure they want to share data with third parties. To be discussed what data will fall into the extraordinary bucket.

**Comment by webcrumbs-community** [comment]
**Performance** is key. There must be best practices guides on the use and sharing of libraries in plugin development.

**Comment by webcrumbs-community** [comment]
On **translation**. It appears that with Gen AI there are more intelligent ways of allowing internalization of plugins than to hard code variables, as it is done in Wordpress. One possible way to go: a (native or not native?) plugin scraps all text content of the website and displays it on the admin panel, as rows, with automatic suggestions by Gen AI for each language column, and allows overwriting the suggestions manually.

**Comment by webcrumbs-community** [comment]
Thread to think of **communication** between plugins and container.

E.g. Should the plugin be able to change pathname? Should the plugin be able to react to pathnames?
